### PR TITLE
[deltatocumulativeprocessor] Introduce an upper bound for exp histogram buckets

### LINKED ISSUE #33277: deltatocumulative: Number of buckets in exponential histograms should be capped
### Component(s)

processor/deltatocumulative

## Description
Currently size of an individual histogram is unbounded and can grow until OOM is reached.
This is prominent if an application has large distribution of data overall but only relatively small in an individual delta datapoint. In this case cumulative aggregation keeps the scale but keeps growing number of buckets to fit all the data.

## Steps to Reproduce
Send two histogram datapoints with the same scale but drastically different offsets.

## Expected Result
Cumulative exponential histogram is downscaled to keep around ≈160 buckets.

## Actual Result
Exponential histogram grows number of buckets instead, leading to OOM.

### Collector version

v0.101.0

## Environment
OS: Linux
Compiler(if manually compiled): go 1.22.2

### OpenTelemetry Collector configuration

_No response_

### Log output

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by euroelessar**
#### Description
This PR introduces a limit on the maximum number of exponential histogram buckets within the deltatocumulativeprocessor. Previously, when merging delta metrics into cumulative metrics, the resulting exponential histograms could grow very large, potentially causing excessive memory usage and processing overhead. By capping the number of buckets at 160 and dynamically downscaling histograms when necessary, this change ensures that the processor remains efficient and stable even when handling large, merged exponential histograms.

#### Link to tracking issue
Fixes #33277

#### Testing
Added unit tests for edge cases.

#### Documentation
Updated changelog.

### CODE REVIEW

**Comment by euroelessar** [comment]
@sh0rez Please have a look, it's an alternative take on #34157, it avoids any extra memory allocations and could be easier to follow.
All changes are in `ExpHistogram.Add` (and `Merge` called from it), as it's the only place which can grow the histogram.

File: processor/deltatocumulativeprocessor/internal/data/add.go [hunk 1]
```diff
@@ -76,6 +106,18 @@ func (dp ExpHistogram) Add(in ExpHistogram) ExpHistogram {
 		hi.SetScale(lo.Scale())
 	}
 
+	// Downscale if an expected number of buckets after the merge is too large.
+	if deltaScale := max(getDeltaScale(dp.Positive(), in.Positive()), getDeltaScale(dp.Negative(), in.Negative())); deltaScale > 0 {
```
Comment (sh0rez) [inline processor/deltatocumulativeprocessor/internal/data/add.go:110]: iiuc, this downscales using the existing algorithm first, possibly exceeding the cap.
we check for that only after the fact, already having allocated more memory than we want, right?
this sounds like it's still vulnerable to DoS attacks

Comment (euroelessar) [reply processor/deltatocumulativeprocessor/internal/data/add.go:110]: Downscaling operation should always produce no more buckets than the original, right?
In particular, `expo.Downscale` never performs allocations (except panicking path), so it can not be a DoS attack vector on memory.

**Comment by euroelessar** [comment]
@sh0rez hi, would you be able to have an another look at the comment above?

File: processor/deltatocumulativeprocessor/internal/data/expo_test.go [hunk 2]
```diff
@@ -85,9 +85,33 @@ func TestExpoAdd(t *testing.T) {
...
 			expotest.ObserveInto(bs, expo.Scale(0), 1, 2, 3, 4)
 			expotest.ObserveInto(bs, expo.Scale(0), 4, 3, 2, 1)
-			bs.BucketCounts().Append([]uint64{0, 0}...) // rescaling leaves zeroed memory. this is expected
 			return bs
 		}()},
+	}, {
+		name: "scale/no_downscale_within_limit",
+		dp:   expdp{Scale: 0, PosNeg: generateBins(10, 80, 1), Count: 160},
+		in:   expdp{Scale: 0, PosNeg: generateBins(80+10, 60, 2), Count: 120},
+		want: expdp{Scale: 0, PosNeg: generateBins(10, 80, 1, 60, 2), Count: 280},
```
Comment (sh0rez) [inline processor/deltatocumulativeprocessor/internal/data/expo_test.go:94]: why do you need `generateBins` and `downscaled`? 
From looking at this test case, it's hard to reason about it. I cannot clearly see what either the input or the output looks like and as such not tell whether it's correct.
Is there a way to stick to written-out `bins{}` literals?

**Review by tombrk — CHANGES_REQUESTED** [review]
sorry for the slow review!

File: processor/deltatocumulativeprocessor/internal/data/add.go [hunk 3]
```diff
@@ -65,6 +67,34 @@ func (dp Histogram) Add(in Histogram) Histogram {
 	return dp
 }
 
+// getDeltaScale computes how many times the histograms need to be downscaled to ensure
+// the bucket range after their merge fits within maxBuckets.
+// This logic assumes that trailing and leading zeros are going to be removed.
+func getDeltaScale(arel, brel pmetric.ExponentialHistogramDataPointBuckets) expo.Scale {
```
[the remark this answers is not in the document]
Comment (euroelessar) [reply processor/deltatocumulativeprocessor/internal/data/add.go]: done

File: processor/deltatocumulativeprocessor/internal/data/expo_test.go (hunk 2 above)
Comment (euroelessar) [reply processor/deltatocumulativeprocessor/internal/data/expo_test.go]: made `maxBuckets` configurable in code (var instead of const), updated the test to set it to max 8 buckets and moved to `rawbs` method
should be easier to read now, please let me know if it needs any further changes!

**Review by tombrk — APPROVED** [review]

**Comment by sh0rez** [comment]
@open-telemetry/collector-contrib-approvers this is ready to merge

**Review by mx-psi — APPROVED** [review]
LGTM and the chosen size aligns with the [SDK default](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#base2-exponential-bucket-histogram-aggregation) :+1:

File: processor/deltatocumulativeprocessor/internal/data/add.go [hunk 4]
```diff
@@ -76,6 +78,21 @@ func (dp ExpHistogram) Add(in ExpHistogram) ExpHistogram {
 		hi.SetScale(lo.Scale())
 	}
 
+	// Downscale if an expected number of buckets after the merge is too large.
+	from := expo.Scale(dp.Scale())
+	to := max(
+		expo.Limit(maxBuckets, from, dp.Positive(), in.Positive()),
+		expo.Limit(maxBuckets, from, dp.Negative(), in.Negative()),
+	)
```
Comment (tiit-clarifai) [inline processor/deltatocumulativeprocessor/internal/data/add.go:86]: @euroelessar @mx-psi Shouldn't this be `min` instead of `max`? What I am seeing is that if I have no negative data points, then no scaling happens and the number of positive buckets still explodes.

I think the tests missed this because they all use `PosNeg`, i.e., use the same positive and negative data points.

Comment (mx-psi) [reply processor/deltatocumulativeprocessor/internal/data/add.go:86]: @tiit-clarifai can you file an issue for this? It definitely looks suspicious

Comment (tiit-clarifai) [reply processor/deltatocumulativeprocessor/internal/data/add.go:86]: Done, https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/37416.

Comment (euroelessar) [reply processor/deltatocumulativeprocessor/internal/data/add.go:86]: My bad, #37432 fixes it. Thanks for letting me know about it! I've also added tests to catch it in the future.
